py.test can't multiple parametrize with ids if it parametrizes class with two or more test methods

Originally reported by: **sergey chipiga (BitBucket: [schipiga](http://bitbucket.org/schipiga), GitHub: [schipiga](http://github.com/schipiga))**

----------------------------------------

Code example

```python
import pytest

@pytest.mark.parametrize('x', [0], ids=['c'])
@pytest.mark.parametrize('y', [0, 1], ids=['a', 'b'])
class Test(object):

    def test1(self, x, y):
        pass

    def test2(self, x, y):
        pass
```

stack trace

```
py.test proba.py --collect-only
============================================================================== test session starts ===============================================================================
platform darwin -- Python 2.7.9 -- py-1.4.26 -- pytest-2.7.0
rootdir: /Users/lira/tmp, inifile: 
collected 0 items / 1 errors 

===================================================================================== ERRORS =====================================================================================
___________________________________________________________________________ ERROR collecting proba.py ____________________________________________________________________________
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/runner.py:149: in __init__
    self.result = func()
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/main.py:437: in _memocollect
    return self._memoizedcall('_collected', lambda: list(self.collect()))
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/main.py:314: in _memoizedcall
    res = function()
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/main.py:437: in <lambda>
    return self._memoizedcall('_collected', lambda: list(self.collect()))
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:563: in collect
    return super(Instance, self).collect()
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:355: in collect
    res = self.makeitem(name, obj)
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:367: in makeitem
    collector=self, name=name, obj=obj)
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/core.py:521: in __call__
    return self._docall(self.methods, kwargs)
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/core.py:528: in _docall
    firstresult=self.firstresult).execute()
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/core.py:393: in execute
    return wrapped_call(method(*args), self.execute)
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/core.py:109: in wrapped_call
    wrap_controller.send(call_outcome)
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:245: in pytest_pycollect_makeitem
    res = list(collector._genfunctions(name, obj))
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:384: in _genfunctions
    self.ihook.pytest_generate_tests.callextra(methods, metafunc=metafunc)
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/core.py:524: in callextra
    return self._docall(self.methods + methods, kwargs)
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/core.py:528: in _docall
    firstresult=self.firstresult).execute()
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/core.py:394: in execute
    res = method(*args)
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:154: in pytest_generate_tests
    metafunc.parametrize(*marker.args, **marker.kwargs)
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:859: in parametrize
    len(argvalues), len(ids)))
E   ValueError: 2 tests specified with 1 ids
```

----------------------------------------
- Bitbucket: https://bitbucket.org/pytest-dev/pytest/issue/751

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,6 +1,10 @@
 2.8.0.dev (compared to 2.7.X)
 -----------------------------
 
+- fix issue751: multiple parametrize with ids bug if it parametrizes class with
+  two or more test methods. Thanks Sergey Chipiga for reporting and Jan
+  Bednarik for PR.
+
 - fix issue768: docstrings found in python modules were not setting up session
   fixtures. Thanks Jason R. Coombs for reporting and Bruno Oliveira for the PR.
 

--- a/_pytest/mark.py
+++ b/_pytest/mark.py
@@ -291,7 +291,7 @@ class MarkInfo:
         #: positional argument list, empty if none specified
         self.args = args
         #: keyword argument dictionary, empty if nothing specified
-        self.kwargs = kwargs
+        self.kwargs = kwargs.copy()
         self._arglist = [(args, kwargs.copy())]
 
     def __repr__(self):

--- a/testing/python/collect.py
+++ b/testing/python/collect.py
@@ -472,6 +472,24 @@ class TestFunction:
         config.pluginmanager.register(MyPlugin2())
         config.hook.pytest_pyfunc_call(pyfuncitem=item)
 
+    def test_issue751_multiple_parametrize_with_ids(self, testdir):
+        modcol = testdir.getmodulecol("""
+            import pytest
+            @pytest.mark.parametrize('x', [0], ids=['c'])
+            @pytest.mark.parametrize('y', [0, 1], ids=['a', 'b'])
+            class Test(object):
+                def test1(self, x, y):
+                    pass
+                def test2(self, x, y):
+                    pass
+        """)
+        colitems = modcol.collect()[0].collect()[0].collect()
+        assert colitems[0].name == 'test1[a-c]'
+        assert colitems[1].name == 'test1[b-c]'
+        assert colitems[2].name == 'test2[a-c]'
+        assert colitems[3].name == 'test2[b-c]'
+
+
 class TestSorting:
     def test_check_equality(self, testdir):
         modcol = testdir.getmodulecol("""

